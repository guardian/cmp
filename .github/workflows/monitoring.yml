name: Monitoring
on:
  push:
    branches:
      - add-cdk

jobs:
  CDK:
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v3

      - name: Setup Node
        uses: actions/setup-node@v3
        with:
          node-version-file: '.nvmrc'

      - name: Install Root dependencies
        run: yarn

      - name: Install CDK dependencies
        working-directory: './monitoring'
        run: yarn

      - name: Test
        working-directory: './monitoring'
        run: yarn test

      - name: Build
        working-directory: './monitoring'
        run: yarn build



