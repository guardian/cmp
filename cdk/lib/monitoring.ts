import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuStack } from '@guardian/cdk/lib/constructs/core';
import { GuLambdaFunction } from '@guardian/cdk/lib/constructs/lambda';
import type { App } from 'aws-cdk-lib';
import { Duration } from 'aws-cdk-lib';
import {
	Alarm,
	ComparisonOperator,
	Metric,
	Unit,
} from 'aws-cdk-lib/aws-cloudwatch';
import { Rule, RuleTargetInput, Schedule } from 'aws-cdk-lib/aws-events';
import { LambdaFunction } from 'aws-cdk-lib/aws-events-targets';
import { Effect, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { Runtime } from 'aws-cdk-lib/aws-lambda';

export class Monitoring extends GuStack {
	constructor(scope: App, id: string, props: GuStackProps) {
		super(scope, id, props);

		const stage = this.stage;

		const region = props.env?.region ?? 'eu-west-1';

		const lambdaBaseName = 'cmp-monitoring';

		const policyStatement = new PolicyStatement({
			effect: Effect.ALLOW,
			actions: ['cloudwatch:PutMetricData'],
			resources: ['*'],
		});

		const monitoringLambdaFunction = new GuLambdaFunction(
			this,
			lambdaBaseName,
			{
				app: `${lambdaBaseName}-lambda-${region}`,
				functionName: `${lambdaBaseName}-${stage}`,
				fileName: `${lambdaBaseName}-lambda-${region}.zip`,
				handler: 'index.handler',
				runtime: Runtime.NODEJS_14_X,
				timeout: Duration.seconds(300),
				memorySize: 2048,
				initialPolicy: [policyStatement],
			},
		);

		new Metric({
			namespace: `Application`,
			metricName: 'CmpLoadingTime',
			period: Duration.minutes(1),
			region: region,
			unit: Unit.SECONDS,
		});

		// Defining metric for lambda errors each minute
		const errorMetric = monitoringLambdaFunction.metricErrors({
			period: Duration.minutes(1),
		});

		// Defining metric for lambda errors each minute
		monitoringLambdaFunction.metricInvocations({
			period: Duration.minutes(1),
		});

		const lambdaEventTarget = new LambdaFunction(monitoringLambdaFunction, {
			event: RuleTargetInput.fromObject({
				stage: 'PROD', // Both scheduled cmp-monitoring-CODE and cmp-monitoring-PROD are monitoring prod versions
				region: region,
			}),
		});

		const monitoringDuration: Duration =
			stage === 'PROD' ? Duration.minutes(2) : Duration.days(1); // Every day for CODE; Every 2 minutes for PROD.

		new Rule(this, 'cmp monitoring schedule', {
			schedule: Schedule.rate(monitoringDuration),
			targets: [lambdaEventTarget],
		});

		// Error Alarm
		new Alarm(this, 'cmp-monitoring-alarms', {
			comparisonOperator: ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
			threshold: 1,
			evaluationPeriods: 1,
			metric: errorMetric,
			alarmDescription:
				'Alarm if the SUM of Errors is greater than or equal to the threshold (1) for 1 evaluation period',
		});
	}
}
