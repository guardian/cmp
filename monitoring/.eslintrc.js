module.exports = {
	root: true,
	env: {
		node: true,
		jest: true,
	},
	extends: [
		'eslint:recommended',
		'plugin:@typescript-eslint/eslint-recommended',
		'plugin:@typescript-eslint/recommended',
		'@guardian/eslint-config-typescript',
	],
	parserOptions: {
		ecmaVersion: 2020,
		tsconfigRootDir: __dirname,
		sourceType: 'module',
		project: ['./tsconfig.eslint.json'],
	},
	rules: {
		'@typescript-eslint/no-inferrable-types': 0,
		'import/no-namespace': 2,
	},
	ignorePatterns: ['node_modules', 'dist'],
};
