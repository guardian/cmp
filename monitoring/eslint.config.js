import globals from "globals";
import path from 'path';
import { fileURLToPath } from 'url';
import js from "@eslint/js"
import prettier from "eslint-config-prettier"
import { FlatCompat } from "@eslint/eslintrc";
//import guardianConfig from "@guardian/eslint-config-typescript"; //Use this once an es-module compatible version is available

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const compat = new FlatCompat({
    baseDirectory: __dirname,
	recommendedConfig: js.configs.recommended,
});

export default [
  ...compat.extends("@guardian/eslint-config-typescript", ''), //replace with guardianConfig once an es-module compatible version is available
 //guardianConfig,
  {
    files: ["**/*.ts"],

	languageOptions: {
		ecmaVersion: 2022,
		sourceType: "module",
		globals: { ...globals.jest, ...globals.node },
		parserOptions: {
			ecmaVersion: 2022,
			tsconfigRootDir: __dirname,
			sourceType: 'module',
			project: ['./tsconfig.eslint.json'],
		}
	},

    plugins: { compat },
	rules: {
		'@typescript-eslint/no-inferrable-types': 0,
		'import/no-namespace': 2,
		'import/newline-after-import': 0,
		'import/no-named-as-default': 0,
		'import/no-named-as-default-member': 0,
	},
	ignores: ['jest.config.js', 'node_modules', 'dist']
  },
  prettier,
]
