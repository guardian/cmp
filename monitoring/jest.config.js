module.exports = {
	testMatch: ['<rootDir>/src/**/*.test.ts'],
	transform: {
		'^.+\\.tsx?$': 'ts-jest',
	},
	setupFilesAfterEnv: ['./jest.setup.js'],
	preset: 'jest-puppeteer',
};
