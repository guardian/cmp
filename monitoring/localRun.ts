import { program } from 'commander';
import { prompt } from 'inquirer';
import type { CustomScheduleEventContent } from './src';

type LocalRunCLIUserInput = {
	env: string;
	jurisdiction: string;
};

type LocalRunCliArguments = {
	env?: string;
	jurisdiction?: string;
};

class NoArgumentsError extends Error {
	constructor(
		message: string = 'Arguments provided are missing or incorrect',
	) {
		super(message);
		this.name = 'NoArgumentsError';
	}
}

const stages: string[] = ['local', 'code', 'prod'];
const jurisdictions: string[] = ['aus', 'ccpa', 'tcfv2'];

function isArgumentValid(args: LocalRunCliArguments): boolean {
	if (!args.env || !args.jurisdiction) {
		return false;
	}

	return (
		stages.includes(args.env) && jurisdictions.includes(args.jurisdiction)
	);
}

async function handleEvent(options: LocalRunCLIUserInput) {
	const { handler } = await import('./src');

	const event: CustomScheduleEventContent = {
		stage: options.env,
		jurisdiction: options.jurisdiction,
	};
	await handler(event);
	process.exit(0);
}

async function argumentBasedCLI() {
	program
		.option('-e, --env [environment]')
		.option('-j, --jurisdiction [environment]')
		.parse();

	const options: LocalRunCLIUserInput = program.opts();
	if (isArgumentValid(options)) {
		await handleEvent(options);
	} else {
		throw new NoArgumentsError();
	}
}

async function interactiveCLI() {
	const questions = [
		{
			type: 'list',
			name: 'env',
			message: 'Which environment would you like to test?',
			choices: stages,
		},
		{
			type: 'list',
			name: 'jurisdiction',
			message: 'Which jurisdiction would you like to test?',
			choices: jurisdictions,
		},
	];

	await prompt(questions).then(async (userInput: LocalRunCLIUserInput) => {
		await handleEvent(userInput);
	});
}

async function main() {
	try {
		await argumentBasedCLI();
	} catch (error) {
		if (error instanceof NoArgumentsError) {
			await interactiveCLI();
		} else {
			process.exit(1);
		}
	}
}

void main();
