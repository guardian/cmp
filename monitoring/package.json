{
	"name": "monitoring",
	"version": "1.0.0",
	"main": "index.js",
	"license": "MIT",
	"scripts": {
		"test": "jest",
		"test:dev": "jest --watch",
		"lint": "eslint src/** localRun.ts --ext .ts --no-error-on-unmatched-pattern",
		"start": "ts-node-dev localRun.ts",
		"remote": "ts-node-dev remoteRun.ts",
		"build": "ncc build src/index.ts -o dist -e aws-sdk"
	},
	"dependencies": {
		"@aws-sdk/client-cloudwatch": "^3.276.0",
		"@types/jest-environment-puppeteer": "^5.0.2",
		"chrome-aws-lambda": "^10.1.0",
		"commander": "^9.3.0",
		"dotenv": "^10.0.0",
		"inquirer": "^8.2.3",
		"puppeteer-core": "^10.2.0",
		"uuid": "^8.3.2"
	},
	"devDependencies": {
		"@aws-sdk/client-lambda": "^3.105.0",
		"@guardian/eslint-config-typescript": "^1.0.2",
		"@guardian/prettier": "^2.0.0",
		"@tsconfig/node14": "^1.0.1",
		"@types/aws-lambda": "^8.10.101",
		"@types/inquirer": "^8.2.1",
		"@types/jest": "^28.1.6",
		"@types/node": "^17.0.41",
		"@types/puppeteer-core": "^5.4.0",
		"@vercel/ncc": "^0.28.6",
		"aws-lambda": "^1.0.7",
		"eslint": "^8.0.0",
		"jest": "^27.3.1",
		"jest-puppeteer": "^6.1.1",
		"markdown-toc": "^1.2.0",
		"prettier": "^2.7.1",
		"puppeteer": "^10.2.0",
		"ts-jest": "^27.0.7",
		"ts-node-dev": "^2.0.0",
		"typescript": "4.3.5"
	},
	"prettier": "@guardian/prettier",
	"overrides": {
		"clone-deep": "4.0.1"
	}
}
