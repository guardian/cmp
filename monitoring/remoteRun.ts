import type { InvokeCommandOutput } from '@aws-sdk/client-lambda';
import { InvokeCommand, LambdaClient } from '@aws-sdk/client-lambda';
import { prompt } from 'inquirer';
import type { CustomScheduleEventContent } from './src';

type RemoteRunCLIUserInput = {
	stage: string;
	regions: string[];
};

const decode = (str: string) => Buffer.from(str, 'base64').toString();

async function invokeInRegion(
	region: string,
	functionName: string,
	stage: string,
): Promise<InvokeCommandOutput> {
	const payload: CustomScheduleEventContent = {
		region: region,
		stage: stage,
	};
	const command = new InvokeCommand({
		FunctionName: functionName,
		LogType: 'Tail',
		Payload: Buffer.from(JSON.stringify(payload)),
	});

	const client = new LambdaClient({ region: region });
	return await client.send(command);
}

function processResult(result: InvokeCommandOutput) {
	const logResult = result.LogResult;

	if (logResult) {
		console.log(decode(logResult));
	}
}

async function main() {
	const regionsToCheck = [
		'us-west-1',
		'eu-west-1',
		'ap-southeast-2',
		'ca-central-1',
	];

	await prompt([
		{
			type: 'list',
			name: 'stage',
			message: 'Which environment would you like to test?',
			choices: ['prod', 'code'],
		},
		{
			type: 'checkbox',
			name: 'regions',
			message: 'Which environment would you like to test?',
			choices: regionsToCheck,
			default: regionsToCheck,
			validate(input: string[]) {
				if (input.length === 0) {
					return 'Please select a region';
				}
				return true;
			},
		},
	])
		.then(async (userInput: RemoteRunCLIUserInput) => {
			const invokeSettledResults = await Promise.allSettled(
				userInput.regions.map((region) =>
					invokeInRegion(
						region,
						'cmp-monitoring-CODE',
						userInput.stage,
					),
				),
			);

			invokeSettledResults.map((result) => {
				if (result.status == 'fulfilled') {
					processResult(result.value);
				} else {
					console.log('------------------------------------------');
					console.log('Failed to get response: ', result.reason);
					console.log('------------------------------------------');
				}
			});

			process.exit(0);
		})
		.catch((error: Error) => {
			process.exit(1);
		});
}

void main();
