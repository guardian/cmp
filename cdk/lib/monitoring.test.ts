import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { Monitoring } from './monitoring';

describe('The Monitoring stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const stack = new Monitoring(app, 'CmpMonitoringStack', {
			stack: 'cmp-monitoring',
			stage: 'PROD',
		});
		expect(Template.fromStack(stack).toJSON()).toMatchSnapshot();
	});
});
