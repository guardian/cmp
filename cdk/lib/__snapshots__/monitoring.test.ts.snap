// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Monitoring stack matches the snapshot 1`] = `
Object {
  "Metadata": Object {
    "gu:cdk:constructs": Array [
      "GuDistributionBucketParameter",
      "GuLambdaFunction",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": Object {
    "DistributionBucketName": Object {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "cmpmonitoring2CD05B94": Object {
      "DependsOn": Array [
        "cmpmonitoringServiceRoleDefaultPolicy19C1C252",
        "cmpmonitoringServiceRole3F7FDCEC",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "cmp-monitoring/PROD/cmp-monitoring-lambda-eu-west-1/cmp-monitoring-lambda-eu-west-1.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "APP": "cmp-monitoring-lambda-eu-west-1",
            "STACK": "cmp-monitoring",
            "STAGE": "PROD",
          },
        },
        "FunctionName": "cmp-monitoring-PROD",
        "Handler": "index.handler",
        "MemorySize": 2048,
        "Role": Object {
          "Fn::GetAtt": Array [
            "cmpmonitoringServiceRole3F7FDCEC",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "cmp-monitoring-lambda-eu-west-1",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/consent-management-platform",
          },
          Object {
            "Key": "Stack",
            "Value": "cmp-monitoring",
          },
          Object {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "cmpmonitoringServiceRole3F7FDCEC": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "cmp-monitoring-lambda-eu-west-1",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/consent-management-platform",
          },
          Object {
            "Key": "Stack",
            "Value": "cmp-monitoring",
          },
          Object {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "cmpmonitoringServiceRoleDefaultPolicy19C1C252": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "cloudwatch:PutMetricData",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "DistributionBucketName",
                      },
                      "/cmp-monitoring/PROD/cmp-monitoring-lambda-eu-west-1/cmp-monitoring-lambda-eu-west-1.zip",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:ssm:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/cmp-monitoring/cmp-monitoring-lambda-eu-west-1",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:ssm:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/cmp-monitoring/cmp-monitoring-lambda-eu-west-1/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "cmpmonitoringServiceRoleDefaultPolicy19C1C252",
        "Roles": Array [
          Object {
            "Ref": "cmpmonitoringServiceRole3F7FDCEC",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "cmpmonitoringalarms68497862": Object {
      "Properties": Object {
        "AlarmDescription": "Alarm if the SUM of Errors is greater than or equal to the threshold (1) for 1 evaluation period",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "FunctionName",
            "Value": Object {
              "Ref": "cmpmonitoring2CD05B94",
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "Errors",
        "Namespace": "AWS/Lambda",
        "Period": 60,
        "Statistic": "Sum",
        "Threshold": 1,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "cmpmonitoringscheduleAllowEventRuleCmpMonitoringStackcmpmonitoring0A19AA24A7F368FB": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "cmpmonitoring2CD05B94",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "cmpmonitoringscheduleFB1DBC53",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "cmpmonitoringscheduleFB1DBC53": Object {
      "Properties": Object {
        "ScheduleExpression": "rate(2 minutes)",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "cmpmonitoring2CD05B94",
                "Arn",
              ],
            },
            "Id": "Target0",
            "Input": "{\\"stage\\":\\"PROD\\",\\"region\\":\\"eu-west-1\\"}",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
  },
}
`;
